staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +34,445 @@ func (s *Source) ServiceNodeDelete(path string) {
-			} else if !gatewayModel && strings.HasSuffix(path, ConsumerNode) {
-				go s.doWatchInstance(path, s.EndpointUpdate, nil)
-			}
-		}
-	default:
+const (
+	providerPathSuffix = "/providers"
+	consumerPathSuffix = "/consumers"
+)
+
+type EventType int
+
+const (
+	EventTypeCreate EventType = iota
+	EventTypeDelete
+)
+
+type ServiceEvent struct {
+	path  string
+	etype EventType
+}
+
+type EndpointWatcherOpts struct {
+	conn                *atomic.Value
+	endpointUpdateFunc  func(providers, consumers []string, serverPath string)
+	serviceDeleteFunc   func(path string)
+	gatewatModel        bool
+	initCallbackFactory func() func()
+}
+
+func NewEndpointWatcher(servicePath string, opts EndpointWatcherOpts) *EndpointWatcher {
+	ew := &EndpointWatcher{
+		conn:         opts.conn,
+		servicePath:  servicePath,
+		gatewatModel: opts.gatewatModel,
+		handler: func(providers, consumers []string) {
+			opts.endpointUpdateFunc(providers, consumers, servicePath)
+		},
+		serviceDeleteHandler: func() {
+			opts.serviceDeleteFunc(servicePath)
+		},
+		signalExit:   make(chan struct{}),
+		exit:         make(chan struct{}),
+		initCallback: opts.initCallbackFactory(),
 	}
+	return ew
 }
 
-func (s *Source) doWatch(path string, updateFunction func([]string, string), deleteFunction func(string)) {
-	if s.watchPath.Has(path) {
-		return
+type EndpointWatcher struct {
+	conn *atomic.Value
+
+	// /{root:dubbo}/{service:<service-name>}
+	// - /{root:dubbo}/{service:<service-name>}/providers/dubbo://ip:port/{provider_service}}?xxx
+	// - /{root:dubbo}/{service:<service-name>}/consumers/consumer://ip[:port]/{consumer_service}?xxx
+	servicePath string
+
+	// update by current status
+	handler func(providers, consumers []string)
+
+	serviceDeleteHandler func()
+
+	gatewatModel bool
+
+	signalExit, exit chan struct{}
+
+	once         sync.Once
+	initCallback func()
+}
+
+// non block
+func (ew *EndpointWatcher) Start(ctx context.Context) {
+	scope.Debugf("zk endpointWatcher %q start watching", ew.servicePath)
+	if !ew.gatewatModel {
+		go ew.watchBoth(ctx)
+	} else {
+		go ew.watchProviderOnly(ctx)
 	}
-	scope.Infof("try to watch the zk path: %q", path)
-	initWatch := true
+}
+
+func (ew *EndpointWatcher) Exit() {
+	close(ew.signalExit)
+	// wait for exit
+	<-ew.exit
+}
+
+func (ew *EndpointWatcher) watchProviderOnly(ctx context.Context) {
+	defer close(ew.exit)
 	for {
-		children, _, e, err := s.Con.ChildrenW(path)
-		if err != nil {
-			if shouldStopWatch(err) {
-				if !initWatch {
-					if deleteFunction != nil {
-						deleteFunction(path)
-					}
-					s.watchPath.Remove(path)
-				}
-				scope.Warnf("stop watch the zk path %q with err: %v", path, err)
-				return
-			}
-			scope.Errorf("watch the zk path %q failed: %v", path, err)
-			time.Sleep(time.Second)
+		select {
+		case <-ew.signalExit:
+			ew.serviceDeleteHandler()
+			scope.Infof("endpointWatcher %q's providerOnlyWatching exit due to service deleted", ew.servicePath)
+			return
+		default:
```

Comment:
mv `case <-eventCh` and `case <-ctx.Done()` here

Reply:
exit has higher priority